near-sdk/src/json_types/public_key.rs

```diff
@@ -77,19 +87,29 @@ impl Base58PublicKey {
                 kind: ParsePublicKeyErrorKind::InvalidLength(data.len()),
             });
         }
+        let mut bytes = Vec::with_capacity(1 + expected_length);
+        bytes.push(curve as u8);
+        bytes.extend(data);
 
-        Ok(Self { curve, data })
+        Ok(Self { data: bytes })
     }
 
     /// Get info about the CurveType for this public key
     pub fn curve_type(&self) -> CurveType {
-        self.curve
+        CurveType::from_u8(self.data[0]).unwrap()
+    }
+}
+
+impl Deref for Base58PublicKey {
+    type Target = Vec<u8>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.data
     }
 }
```

Comment:
This impl can be difficult to maintain, as it removes any flexibility to change the underlying type later. We need to be sure this will be the final type if doing this. `Deref` impls can be very tricky to maintain in a backwards compatible way

Reply:
That's a good point. Maybe `as_vec() -> &Vec<u8>` would be better for now, even though this would add on to the list of breaking changes

Reply:
That function would be just as hard to maintain, because it still requires the underlying type to be `Vec` to reference it. Could expose as_bytes with `&[u8]` to allow `Box<[u8]>` or some type that allows partial stack allocation.

Also pretty sure clippy warns against `&Vec<u8>` because there is no benefit to this

Reply:
I'd say `From<PublicKey> for Vec<u8>`, `PublicKey::into_bytes`  and `PublicKey::as_bytes` would be the right API here (that's how `std::String` exposes its internal impl)